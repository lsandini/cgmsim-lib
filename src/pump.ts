import logger, { getDeltaMinutes, getTreatmentActivity } from './utils';
import * as moment from 'moment';
import { Profile, Treatment } from './Types';

function getProfileSwitch(treatments, duration) {
	const computedProfilesSwitch: {
		start: moment.Moment;
		end: moment.Moment;
		basal: { time: string; value: number }[] | number;
	}[] = [];
	const now = moment().utc();
	treatments
		.filter(
			(e) =>
				e.created_at &&
				now.diff(moment(e.created_at), 'minutes') <= duration && // temps basals set in the last 3 hours
				e.eventType === 'Profile Switch' &&
				e.duration != 0
		)
		.sort((f, s) => {
			return moment(f.created_at).diff(s.created_at);
		})
		.forEach((tr) => {
			if (tr.profileJson && tr.percentage) {
				const start = moment(tr.created_at).utc();
				const end = moment(tr.created_at).add(tr.duration, 'minutes').utc();
				const profile = JSON.parse(tr.profileJson);
				if (Array.isArray(profile.basal)) {
					const basals = profile.basal.map((pb) => ({
						time: pb.time,
						value: pb.value * (tr.percentage / 100),
					}));
					computedProfilesSwitch.push({
						start,
						basal: basals,
						end,
					});
				} else {
					computedProfilesSwitch.push({
						start,
						basal: profile.basal * (tr.percentage / 100),
						end,
					});
				}
			} else {
				const currentIndex = computedProfilesSwitch.length - 1;
				if (currentIndex >= 0) {
					computedProfilesSwitch[currentIndex].end = moment(tr.created_at);
				}
			}
		});
	return computedProfilesSwitch;
}

function getTempBasal(treatments, duration) {
	const computedTempBasal: {
		start: moment.Moment;
		end: moment.Moment;
		insulin: number;
	}[] = [];
	const now = moment().utc();
	treatments
		.filter(
			(e) =>
				e.created_at &&
				now.diff(moment(e.created_at), 'minutes') <= duration && // temps basals set in the last 3 hours
				e.eventType === 'Temp Basal' &&
				e.duration != 0
		)
		.sort((f, s) => {
			return moment(f.created_at).diff(s.created_at);
		})
		.forEach((b) => {
			if (b.absolute !== undefined) {
				const start = moment(b.created_at).utc();
				const end = moment(b.created_at).add(b.duration, 'minutes').utc();
				computedTempBasal.push({
					start,
					insulin: b.absolute,
					end,
				});
			} else {
				const currentIndex = computedTempBasal.length - 1;
				if (currentIndex >= 0) {
					computedTempBasal[currentIndex].end = moment(b.created_at);
				}
			}
		});
	return computedTempBasal;
}

function getBasalFromProfiles(
	orderedProfiles: Profile[],
	currentAction: moment.Moment
) {
	//last basal before the end
	const activeProfiles = orderedProfiles.filter(
		(p) => moment(p.startDate).diff(currentAction) <= 0
	);
	if (activeProfiles && activeProfiles.length > 0) {
		const activeProfile = activeProfiles[0];
		const activeProfileName = activeProfile.defaultProfile;
		const activeProfileBasals = activeProfile.store[activeProfileName].basal;
		return activeBasalByTime(activeProfileBasals, currentAction);
	}
	return 0;
}

function activeBasalByTime(
	activeProfileBasals:
		| { value: number; time: string; timeAsSecond?: number }[]
		| number,
	currentAction: moment.Moment
) {
	if (Array.isArray(activeProfileBasals)) {
		const compatiblesBasalProfiles = activeProfileBasals.filter((b) => {
			return b.time.localeCompare(currentAction.format('HH:mm')) <= 0;
		});
		const index = compatiblesBasalProfiles.length - 1;
		const currentBasal = compatiblesBasalProfiles[index];
		return currentBasal.value;
	} else {
		return activeProfileBasals;
	}
}

export default function (
	treatments: Treatment[],
	profiles: Profile[],
	dia: number,
	peak: number
) {
	const basalAsBoluses: { minutesAgo: number; insulin: number }[] = [];
	const endDiaAction = moment().utc();
	const startDiaAction = moment()
		.add(-dia, 'hour')
		.set({ minute: 0, second: 0, millisecond: 0 })
		.utc();
	const duration = dia * 60;

	const orderedProfiles = profiles
		.filter((profile) => {
			const profileName = profile.defaultProfile;
			return profile.store[profileName];
		})
		.sort((first, second) =>
			moment(second.startDate).diff(moment(first.startDate))
		);

	const computedTempBasal = getTempBasal(treatments, duration);
	const computedProfileSwitch = getProfileSwitch(treatments, duration);

	// const basalsToUpdate = [];
	for (
		let currentAction = startDiaAction;
		currentAction.diff(endDiaAction) <= 0;
		currentAction.add(5, 'minutes')
	) {
		const tempBasalActives = computedTempBasal.filter(
			(t) => t.start.diff(currentAction) <= 0 && t.end.diff(currentAction) > 0
		);
		const profilesStichActives = computedProfileSwitch.filter(
			(t) => t.start.diff(currentAction) <= 0 && t.end.diff(currentAction) > 0
		);
		let basalToUpdate: { minutesAgo: number; insulin: number };
		//if there is a temp basal actives
		if (tempBasalActives.length > 0) {
			basalToUpdate = {
				minutesAgo: getDeltaMinutes(currentAction.toISOString()),
				insulin: tempBasalActives[0].insulin / 12,
			};
		} else if (profilesStichActives.length > 0) {
			const basal = activeBasalByTime(
				profilesStichActives[0].basal,
				currentAction
			);
			basalToUpdate = {
				minutesAgo: getDeltaMinutes(currentAction.toISOString()),
				insulin: basal / 12,
			};
		} else {
			let currentBasal = {
				value: getBasalFromProfiles(orderedProfiles, currentAction),
			};
			basalToUpdate = {
				minutesAgo: getDeltaMinutes(currentAction.toISOString()),
				insulin: currentBasal.value / 12,
			};
		}
		basalAsBoluses.push(basalToUpdate);
	}

	const pumpBasalAct = basalAsBoluses.reduce(
		(tot, entry) =>
			tot + getTreatmentActivity(peak, duration, entry.minutesAgo, entry.insulin),
		0
	);
	logger.debug("the pump's basal activity is: %o", pumpBasalAct);
	return pumpBasalAct;
}
