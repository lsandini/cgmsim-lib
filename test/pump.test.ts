import moment = require('moment');
import pump from '../src/pump';

const dia = 6;
const peak = 90;

const dynamicBasal = [
  {
    time: '00:00',
    value: 1,
  },
  {
    time: '12:00',
    value: 1.2,
  },
  {
    time: '14:00',
    value: 0.8,
  },
  {
    time: '18:00',
    value: 1,
  },
];
const profileSwitch = [
  {
    eventType: 'Profile Switch',
    percentage: 130,
    duration: 30,
    created_at: '2022-05-06T13:00:00.000Z',
    profileJson: JSON.stringify({
      basal: dynamicBasal,
    }),
  },
];
const tmpBasal = [
  {
    eventType: 'Temp Basal',
    absolute: 0,
    duration: 60,
    created_at: '2022-05-06T11:00:00.000Z',
  },
  {
    eventType: 'Temp Basal',
    absolute: 1.2,
    duration: 30,
    created_at: '2022-05-06T16:00:00.000Z',
  },
];
describe('test pump', () => {
  const date = new Date('2022-05-07T11:20:00Z');

  beforeEach(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(date);
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  test('no profile', () => {
    let _date = moment('2022-05-06T16:30:00.000Z');
    jest.setSystemTime(_date.toDate());

    const result = pump([], [], dia, peak);

    expect(result).toBe(0);
  });
  test('static basal in profile and no treatments', () => {
    let _date = moment('2022-05-06T16:30:00.000Z');
    jest.setSystemTime(_date.toDate());

    const result = pump(
      [],
      [
        {
          startDate: '2000-01-01T00:00:00.000Z',
          defaultProfile: 'Default',
          store: {
            Default: {
              basal: 0.7,
            },
          },
        },
      ],
      dia,
      peak
    );

    expect(result).toMatchSnapshot();
  });
  test('static basal in profile and temp basal 0', () => {
    let _date = moment('2022-05-06T16:30:00.000Z');
    jest.setSystemTime(_date.toDate());
    const treatments = [
      {
        eventType: 'Temp Basal',
        absolute: 0,
        duration: 600,
        created_at: '2022-05-06T10:31:00.000Z',
      },
    ];
    const result = pump(
      treatments,
      [
        {
          startDate: '2000-01-01T00:00:00.000Z',
          defaultProfile: 'Default',
          store: {
            Default: {
              basal: 0.7,
            },
          },
        },
      ],
      dia,
      peak
    );

    expect(result).toBe(0);
  });

  test('static basal in profile and 2 different temp basal', () => {
    let _date = moment('2022-05-06T16:30:00.000Z');
    jest.setSystemTime(_date.toDate());
    const treatments = tmpBasal;
    const result = pump(
      treatments,
      [
        {
          startDate: '2000-01-01T00:00:00.000Z',
          defaultProfile: 'Default',
          store: {
            Default: {
              basal: 0.7,
            },
          },
        },
      ],
      dia,
      peak
    );

    expect(result).toMatchSnapshot();
  });
  test('dynamic  basal in profile and no temp basal', () => {
    let _date = moment('2022-05-06T16:30:00.000Z');
    jest.setSystemTime(_date.toDate());
    const treatments = [];
    const result = pump(
      treatments,
      [
        {
          startDate: '2000-01-01T00:00:00.000Z',
          defaultProfile: 'Default',
          store: {
            Default: {
              basal: dynamicBasal,
            },
          },
        },
      ],
      dia,
      peak
    );

    expect(result).toMatchSnapshot();
  });
  test('dynamic  basal in profile, no temp basal and profile switch', () => {
    let _date = moment('2022-05-06T16:30:00.000Z');
    jest.setSystemTime(_date.toDate());
    const treatments = profileSwitch;
    const result = pump(
      treatments,
      [
        {
          startDate: '2000-01-01T00:00:00.000Z',
          defaultProfile: 'Default',
          store: {
            Default: {
              basal: dynamicBasal,
            },
          },
        },
      ],
      dia,
      peak
    );

    expect(result).toMatchSnapshot();
  });
  test('dynamic basal in profile and temp basal', () => {
    let _date = moment('2022-05-06T16:30:00.000Z');
    jest.setSystemTime(_date.toDate());
    const treatments = tmpBasal;

    const result = pump(
      treatments,
      [
        {
          startDate: '2000-01-01T00:00:00.000Z',
          defaultProfile: 'Default',
          store: {
            Default: {
              basal: dynamicBasal,
            },
          },
        },
      ],
      dia,
      peak
    );

    expect(result).toMatchSnapshot();
  });
});
