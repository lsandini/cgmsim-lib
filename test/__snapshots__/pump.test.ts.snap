// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test pump dynamic  basal in profile and no temp basal 1`] = `0.01559265706624232`;

exports[`test pump dynamic  basal in profile, no temp basal and profile switch 1`] = `0.01614147760949147`;

exports[`test pump dynamic basal in profile and temp basal 1`] = `0.015201408758962516`;

exports[`test pump static basal in profile and 2 different temp basal 1`] = `0.011547188344800504`;

exports[`test pump static basal in profile and no treatments 1`] = `0.011609624381664367`;
